File: mobile/src/main/kotlin/de/scout/fireplace/home/GalleryAdapter.java
package de.scout.fireplace.home;

import android.support.v4.view.PagerAdapter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import butterknife.ButterKnife;
import com.squareup.picasso.Picasso;
import de.scout.fireplace.R;
import de.scout.fireplace.models.Expose;
import java.util.List;

public class GalleryAdapter extends PagerAdapter {

  private final List<Expose.Picture> pictures;

  GalleryAdapter(List<Expose.Picture> pictures) {
    this.pictures = pictures;
  }

  @Override
  public int getCount() {
    return pictures.size();
  }

  @Override
  public boolean isViewFromObject(View view, Object object) {
    return view == object;
  }

  @Override
  public Object instantiateItem(ViewGroup container, int position) {
    View view = LayoutInflater.from(container.getContext()).inflate(R.layout.item_gallery, container, false);
    View parent = (View) container.getParent();

    Picasso.with(view.getContext())
        .load(pictures.get(position).getUrl(parent.getWidth(), parent.getHeight()))
        .fit()
        .centerCrop()
        .into(ButterKnife.<ImageView>findById(view, R.id.image));

    container.addView(view);
    return view;
  }

  @Override
  public void destroyItem(ViewGroup container, int position, Object object) {
    container.removeView((View) object);
  }
}

File: mobile/src/main/kotlin/de/scout/fireplace/home/GalleryAdapter.kt
package de.scout.fireplace.home

import android.support.v4.view.PagerAdapter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import butterknife.ButterKnife
import com.squareup.picasso.Picasso
import de.scout.fireplace.R
import de.scout.fireplace.models.Expose

class GalleryAdapter internal constructor(private val pictures: List<Expose.Picture>, private val callback: (Int) -> Unit) : PagerAdapter() {

  override fun getCount(): Int {
    return pictures.size
  }

  override fun isViewFromObject(view: View, obj: Any): Boolean {
    return view === obj
  }

  override fun instantiateItem(container: ViewGroup, position: Int): Any {
    val view = LayoutInflater.from(container.context).inflate(R.layout.item_gallery, container, false)
    val parent = container.parent as View

    Picasso.with(view.context)
        .load(pictures.first().getUrl(parent.width, parent.height))
        .fit()
        .centerCrop()
        .into(ButterKnife.findById<ImageView>(view, R.id.image))

    callback.invoke(position)
    container.addView(view)
    return view
  }

  private fun <T> List<T>.first(): T {
    return get(0)
  }

  override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
    container.removeView(`object` as View)
  }
}
